### ISSUE TITLE
Backport Darwin TBD Fixes to 20.09

### ISSUE THREAD

**Issue by TravisWhitaker** [NixOS/nixpkgs#118103]
The tip of `nixos-20.09` isn't usable on MacOS Big Sur because it's missing the new TBD-oriented stdenv:

```
$ nix-build -A hello --check
checking outputs of '/nix/store/19nkhsryfvzp66fkl1mf5026xj7gxdqm-hello-2.10.drv'...
unpacking sources
unpacking source archive /nix/store/3x7dwzq014bblazs7kq20p9hyzz0qh8g-hello-2.10.tar.gz
source root is hello-2.10
setting SOURCE_DATE_EPOCH to timestamp 1416139241 of file hello-2.10/ChangeLog
patching sources
configuring
configure flags: --disable-dependency-tracking --prefix=/nix/store/180066650sp74bhy3fxl4234z8vdnhnz-hello-2.10
checking for a BSD-compatible install... /nix/store/l5y1dp8cbsjysvzvivrac5jrk2bpdh4s-coreutils-8.31/bin/install -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... /nix/store/l5y1dp8cbsjysvzvivrac5jrk2bpdh4s-coreutils-8.31/bin/mkdir -p
checking for gawk... gawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking for gcc... clang
checking whether the C compiler works... no
configure: error: in `/private/tmp/nix-build-hello-2.10.drv-0/hello-2.10':
configure: error: C compiler cannot create executables
See `config.log' for more details
builder for '/nix/store/19nkhsryfvzp66fkl1mf5026xj7gxdqm-hello-2.10.drv' failed with exit code 77
error: build of '/nix/store/19nkhsryfvzp66fkl1mf5026xj7gxdqm-hello-2.10.drv' failed
```

Is there any interest in backporting the TBD stdenv to the 20.09 branch?

**Comment by TravisWhitaker** [comment]
@thefloweringash @LnL7 @matthewbauer do you have any thoughts on how tough this would be to accomplish? Does this PR (https://github.com/NixOS/nixpkgs/pull/107013) represents the current good state of the Darwin stdenv?

**Comment by Gabriel439** [comment]
My understanding is that the key pull request that needs to be backported is this one:

https://github.com/NixOS/nixpkgs/pull/98541

However, that pull request cannot be easily backported to 20.09 because it depends on a few other changes that were made post-20.09:

<details>
<summary>Example patch attempt</summary>

```bash
$ patch -p1 < 98541.patch 
patching file pkgs/stdenv/darwin/default.nix
patching file pkgs/stdenv/darwin/default.nix
Hunk #4 FAILED at 217.
Hunk #5 succeeded at 240 (offset 9 lines).
Hunk #6 succeeded at 256 (offset 9 lines).
Hunk #7 succeeded at 279 (offset 9 lines).
Hunk #8 succeeded at 329 (offset 9 lines).
1 out of 8 hunks FAILED -- saving rejects to file pkgs/stdenv/darwin/default.nix.rej
patching file pkgs/development/compilers/llvm/7/libc++abi.nix
patching file pkgs/stdenv/darwin/default.nix
Hunk #5 FAILED at 246.
Hunk #6 succeeded at 287 (offset 3 lines).
Hunk #7 succeeded at 319 (offset 3 lines).
Hunk #8 succeeded at 370 (offset 3 lines).
1 out of 8 hunks FAILED -- saving rejects to file pkgs/stdenv/darwin/default.nix.rej
patching file pkgs/development/compilers/llvm/10/compiler-rt.nix
patching file pkgs/development/compilers/llvm/5/compiler-rt.nix
patching file pkgs/development/compilers/llvm/6/compiler-rt.nix
patching file pkgs/development/compilers/llvm/7/compiler-rt.nix
patching file pkgs/development/compilers/llvm/8/compiler-rt.nix
patching file pkgs/development/compilers/llvm/9/compiler-rt.nix
patching file pkgs/os-specific/darwin/darwin-stubs/default.nix
patching file pkgs/stdenv/darwin/default.nix
Hunk #1 succeeded at 295 (offset -6 lines).
Hunk #2 succeeded at 340 (offset -6 lines).
Hunk #3 succeeded at 399 with fuzz 2 (offset -6 lines).
Hunk #4 succeeded at 499 (offset -5 lines).
patching file pkgs/top-level/darwin-packages.nix
patching file pkgs/os-specific/darwin/apple-sdk/print-reexports/default.nix
patching file pkgs/os-specific/darwin/apple-sdk/print-reexports/main.c
patching file pkgs/top-level/darwin-packages.nix
patching file pkgs/os-specific/darwin/apple-source-releases/Libsystem/default.nix
patching file pkgs/os-specific/darwin/apple-source-releases/Security/boot.nix
patching file pkgs/os-specific/darwin/apple-source-releases/IOKit/default.nix
patching file pkgs/os-specific/darwin/apple-source-releases/objc4/default.nix
patching file pkgs/os-specific/darwin/apple-sdk/cf-setup-hook.sh
patching file pkgs/os-specific/darwin/apple-sdk/default.nix
patching file pkgs/os-specific/darwin/apple-sdk/frameworks.nix
patching file pkgs/top-level/darwin-packages.nix
patching file pkgs/development/libraries/qt-5/5.12/default.nix
patching file pkgs/development/libraries/qt-5/5.12/qtbase.patch.d/0015-qtbase-tbd-frameworks.patch
patching file pkgs/development/libraries/qt-5/5.14/default.nix
Hunk #1 FAILED at 55.
1 out of 1 hunk FAILED -- saving rejects to file pkgs/development/libraries/qt-5/5.14/default.nix.rej
patching file pkgs/development/libraries/qt-5/5.14/qtbase.patch.d/0012-qtbase-tbd-frameworks.patch
patching file pkgs/development/libraries/qt-5/5.15/default.nix
Hunk #1 FAILED at 55.
1 out of 1 hunk FAILED -- saving rejects to file pkgs/development/libraries/qt-5/5.15/default.nix.rej
patching file pkgs/development/libraries/qt-5/5.15/qtbase.patch.d/0012-qtbase-tbd-frameworks.patch
patching file pkgs/stdenv/darwin/make-bootstrap-tools.nix
Hunk #2 succeeded at 203 (offset -2 lines).
Hunk #3 succeeded at 271 (offset -2 lines).
patching file pkgs/tools/misc/trash-cli/default.nix
patching file pkgs/os-specific/darwin/apple-sdk/default.nix
patching file pkgs/os-specific/darwin/apple-sdk/print-reexports/main.c
```

</details>

**Comment by Gabriel439** [comment]
Actually, this appears to be easier than I thought.  I have a PR to backport the Big Sur fixes to 20.09 here: https://github.com/NixOS/nixpkgs/pull/121040
